File: powerpro/controllers/todo.py
import frappe
from erpnext.assets.doctype.asset_maintenance.asset_maintenance import calculate_next_due_date


@frappe.whitelist()
def close_maintenance_task(todo, task, due_date):
    todo = get_todo(todo)

    if not todo.actions_performed:
        frappe.throw(
            "Debe especificar las acciones realizadas durante esta tarea para poder cerrar el mantenimiento."
        )
    todo.update({
        "status": "Closed",
    })
    todo.flags.ignore_permissions = True
    todo.save()

    maintenance_log = get_asset_maintenance_log(task, due_date)
    maintenance_log.update({
        "completion_date": frappe.utils.nowdate(),
        "actions_performed": todo.actions_performed,
        "completed_by": frappe.session.user,
        "completed_by_name": frappe.get_value("User", frappe.session.user, "full_name"),
        "maintenance_status": "Completed",
    })
    maintenance_log.flags.ignore_permissions = True
    maintenance_log.flags.ignore_validate_update_after_submit = True
    maintenance_log.save()

    if not todo.is_old_maintenance:
        create_next_maintenance_log(maintenance_log)


def create_next_maintenance_log(log):
    asset_maintenance = get_asset_maintenance(log.asset_maintenance)

    due_date = calculate_next_due_date(
        periodicity=log.periodicity,
        start_date=log.due_date,
    )

    asset_maintenance = get_asset_maintenance(log.asset_maintenance)

    for task in asset_maintenance.asset_maintenance_tasks:
        if task.name == log.task:
            task.update({
                "next_due_date": due_date,
                "maintenance_status": "Planned",
            })

    asset_maintenance.flags.ignore_permissions = True
    asset_maintenance.save()


@frappe.whitelist()
def reopen_maintenance_task(todo, task, due_date):
    todo = get_todo(todo)
    todo.update({
        "status": "Open",
        "is_old_maintenance": True,
    })
    todo.flags.ignore_permissions = True
    todo.save()

    log = get_completed_asset_maintenance_log(todo.asset_maintenance_log)
    log.update({
        "completion_date": None,
        "actions_performed": None,
        "maintenance_status": "Planned",
        "completed_by": None,
    })
    log.flags.ignore_permissions = True
    log.db_update()



def get_todo(todo):
    doctype = "ToDo"
    return frappe.get_doc(doctype, todo)


def get_asset_maintenance_log(task, due_date):
    doctype = "Asset Maintenance Log"
    filters = {
        "task": task,
        "due_date": due_date,
        "maintenance_status": ("in", ["Planned", "Overdue"]),
    }
    return frappe.get_doc(doctype, filters)


def get_completed_asset_maintenance_log(log):
    doctype = "Asset Maintenance Log"

    if name := frappe.db.exists(doctype, log):
        return frappe.get_doc(doctype, name)


def get_asset_maintenance(asset_maintenance):
    doctype = "Asset Maintenance"
    return frappe.get_doc(doctype, asset_maintenance)

File: powerpro/public/js/todo.js


frappe.ui.form.off("ToDo", "refresh");

frappe.ui.form.on("ToDo", {
    refresh: function (frm) {
        const { doc } = frm;

        if (doc.reference_type != "Asset Maintenance Task" && doc.reference_name) {
            frm.add_custom_button(__(doc.reference_name), function () {
                frappe.set_route("Form", doc.reference_type, doc.reference_name);
            });
        }

        if (!doc.__islocal) {
            const allowed_doctypes = [
                "Asset Maintenance Task",
                "Training Event",
            ];

            if (doc.status == "Open" && allowed_doctypes.includes(doc.reference_type)) {
                frm.add_custom_button(
                    __("Close"),
                    function () {
                        if (doc.reference_type == "Asset Maintenance Task") {
                            if (!doc.actions_performed) {
                                frappe.throw(
                                    "Debe especificar las acciones realizadas durante esta tarea para poder cerrar el mantenimiento."
                                )
                            }

                            if (doc.__unsaved) {
                                frappe.throw(
                                    "Por favor guarde el registro antes de cerrar la tarea."
                                )
                            }
                                
                            const method = "powerpro.controllers.todo.close_maintenance_task";
                            const args = {
                                "todo": doc.name, 
                                "task": doc.reference_name,
                                "due_date": doc.date,
                            };
                            const callback = function () {
                                frm.reload_doc();
                            }

                            frappe.call({ method, args, callback });
                        } else if (doc.reference_type == "Training Event") {
                            frappe.confirm(
                                __("¿Está seguro que desea cerrar y firmar su asistencia a este evento?"),
                                function () {
                                    frm.set_value("status", "Closed");
                                    frm.save(null, function () {
                                        // back to list
                                        frappe.set_route("List", "ToDo");
                                    });
                                }
                            );
                        } 
                        else {
                            frm.set_value("status", "Closed");
                            frm.save(null, function () {
                                // back to list
                                frappe.set_route("List", "ToDo");
                            });
                        }
                    },
                    "fa fa-check",
                    "btn-success"
                );
            } else {
                if (doc.status == "Cancelled" && doc.reference_type == "Asset Maintenance Task") {
                    return;
                }

                frm.add_custom_button(
                    __("Reopen"),
                    function () {
                        if (doc.reference_type == "Asset Maintenance Task") {
                            const method = "powerpro.controllers.todo.reopen_maintenance_task";
                            const args = {
                                "todo": doc.name, 
                                "task": doc.reference_name,
                                "due_date": doc.date,
                            };
                            const callback = function () {
                                // frm.set_value("status", "Open");
                                frm.reload_doc();
                            }

                            frappe.call({ method, args, callback });
                        }
                        // frm.save();
                    },
                    null,
                    "btn-default"
                );
            }
            // frm.add_custom_button(
            //     __("New"),
            //     function () {
            //         frappe.new_doc("ToDo");
            //     },
            //     null,
            //     "btn-default"
            // );
        }
    },
});